compiler/src/dotty/tools/dotc/transform/patmat/Space.scala

```diff
@@ -291,6 +294,16 @@ trait SpaceLogic {
 class SpaceEngine(implicit ctx: Context) extends SpaceLogic {
   import tpd._
 
+  override def intersectAtomTypes(tp1: Type, tp2: Type) = {
+    def isOpen(tp: Type) =
+      !tp.classSymbol.is(Sealed | Final) && !tp.termSymbol.is(Module)
+
+    if ((tp1.classSymbol.is(Trait) || tp2.classSymbol.is(Trait))
+        && isOpen(tp1) && isOpen(tp2))
+      Typ(AndType(tp1, tp2), true)
```

Comment:
What I've in mind is the case where the two types are `Trait1 & Trait2` and `Trait3 & Trait4`, in which case, the intersection would be `Trait1 & Trait2 & Trait3 & Trait4`.